### PR TITLE
Add Pearson correlation for sort groupby (python)

### PR DESCRIPTION
**Pull request by skirui-source**
Fixes: https://github.com/rapidsai/cudf/issues/8691

### CODE REVIEW

File: python/cudf/cudf/tests/test_dataframe.py [hunk 1]
```diff
@@ -8932,3 +8932,35 @@ def test_frame_series_where_other(data):
...
+        {"id": ["a", "a", "b", "b", "c", "c"], "val": [10, 3, 4, 2, -3, 9]},
+        {
+            "id": [0] * 4 + [1] * 3,
+            "a": [10, 3, 4, 2, -3, 9, 10],
+            "b": [10, 23, -4, 2, -3, 9, 19],
+            "c": [10, -23, -4, 21, -3, 19, 19],
+        },
+    ],
+)
+def test_dataframe_pearson_corr(data):
```
Comment (karthikeyann) [inline python/cudf/cudf/tests/test_dataframe.py:8960]: More tests are needed covering all cases.
- empty/single/multiple column, [0,1,n] elements, [0,1,n] group, same dtypes, mixed dtypes, number of columns, multi-index.
- variations of min_periods.
- error tests (unsupported method, min_period, dtype)

File: python/cudf/cudf/core/groupby/groupby.py [hunk 2]
```diff
@@ -780,6 +783,107 @@ def median(self):
         """Get the column-wise median of the values in each group."""
         return self.agg("median")
 
+    def corr(self, method="pearson", min_periods=1):
+        """
+        Compute pairwise correlation of columns, excluding NA/null values.
+
+        Parameters
+        ----------
+        method: Method of correlation, default 'Pearson'
```
Comment (karthikeyann) [inline python/cudf/cudf/core/groupby/groupby.py:792]:
```suggestion
        method: Method of correlation
```

File: python/cudf/cudf/core/groupby/groupby.py [hunk 3]
```diff
@@ -780,6 +781,105 @@ def median(self):
...
+            Pearson: standard correlation coefficient.
+            Kendall, Spearman correlation and callable method
+            not yet supported.
+
+        min_periods: int, default 1
```
[the remark this answers is not in the document]
Comment (karthikeyann) [reply python/cudf/cudf/core/groupby/groupby.py:797]:
```suggestion
        min_periods: int, optional
```

**Review by karthikeyann — CHANGES_REQUESTED** [review]
more pytest cases.

File: python/cudf/cudf/_lib/cpp/aggregation.pxd [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # Copyright (c) 2020, NVIDIA CORPORATION.
```
Comment (mayankanand007) [inline python/cudf/cudf/_lib/cpp/aggregation.pxd:1]: fix copyright year here as well?

File: python/cudf/cudf/core/groupby/groupby.py [hunk 5]
```diff
@@ -780,6 +784,130 @@ def median(self):
         """Get the column-wise median of the values in each group."""
         return self.agg("median")
 
+    def corr(self, method="pearson", min_periods=1):
```
Comment (mayankanand007) [inline python/cudf/cudf/core/groupby/groupby.py:787]: can you also expose this via the docs by adding it in this file: https://github.com/rapidsai/cudf/blob/branch-21.12/docs/cudf/source/api_docs/groupby.rst?

**Review by mayankanand007 — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/core/groupby/groupby.py [hunk 6]
```diff
@@ -780,6 +784,130 @@ def median(self):
...
+                raise
+
+        # ensure that column-pair labels are arranged in ascending order
+        cols_list = []
+        for i, x in enumerate(_cols):
+            for j, y in enumerate(_cols):
+                if i > j:
+                    cols_list.append((_cols[j], _cols[i]))
+                else:
+                    cols_list.append((_cols[i], _cols[j]))
```
Comment (karthikeyann) [inline python/cudf/cudf/core/groupby/groupby.py:883]: This could use List comprehension too.

File: python/cudf/cudf/core/groupby/groupby.py [hunk 7]
```diff
@@ -780,6 +784,130 @@ def median(self):
...
+        for i, x in zip(cols_split, _cols):
+            ic = gb_corr.loc[:, i].interleave_columns()
+            res[x] = ic
+
+        # create a multiindex for the groupby correlated dataframe,
+        # to match pandas behavior
+        _idx = gb_corr.index.repeat(len(_cols))
+        idx_sort_order = _idx._get_sorted_inds()
+        _idx = _idx._gather(idx_sort_order)
+        # breakpoint()
```
Comment (karthikeyann) [inline python/cudf/cudf/core/groupby/groupby.py:901]:
```suggestion

```

**Review by karthikeyann — CHANGES_REQUESTED** [review]
Looks good.
Just minor cleanups.

File: python/cudf/cudf/core/groupby/groupby.py [hunk 8]
```diff
@@ -780,6 +784,130 @@ def median(self):
...
+        for i in tuple(itertools.combinations_with_replacement(_cols, 2)):
+            new_df._data[i] = cudf.DataFrame._from_data(
+                {"x": self.obj._data[i[0]], "y": self.obj._data[i[1]]}
+            ).to_struct()
+        new_gb = new_df.groupby(by=self._by, level=self._level)
+        try:
+
+            gb_corr = new_gb.agg(lambda x: x.corr(method, min_periods))
+        except RuntimeError as e:
+            if "Unsupported type-agg combination" in str(e):
```
Comment (karthikeyann) [inline python/cudf/cudf/core/groupby/groupby.py:869]: One of the test fails with "Unsupported groupby reduction type-agg combination"

**Comment by karthikeyann** [comment]
rerun tests

**Review by karthikeyann — APPROVED** [review]
LGTM 👍 
Cython looks good. 
Get another python approval.

**Comment by vyasr** [comment]
rerun tests

File: docs/cudf/source/basics/groupby.rst [hunk 9]
```diff
@@ -128,6 +128,30 @@ Aggregations on groups is supported via the ``agg`` method:
     1   4   1  2.0
     2   5   2  4.5
 
+    >>> gdf
```
Comment (vyasr) [inline docs/cudf/source/basics/groupby.rst:131]: I'm a little confused. Why did you need to introduce a new DataFrame in this example? Couldn't you just compute the Pearson correlation from the df that's already here?

File: python/cudf/cudf/_lib/aggregation.pxd [hunk 10]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2020, NVIDIA CORPORATION.
+# Copyright (c) 2020-2021, NVIDIA CORPORATION.
```
[the remark this answers is not in the document]
Comment (vyasr) [reply python/cudf/cudf/_lib/aggregation.pxd:1]: I'm confused, why are we changing this copyright if this file didn't actually change? I think @karthikeyann's suggestion is correct.

File: python/cudf/cudf/_lib/aggregation.pyx [hunk 11]
```diff
@@ -30,6 +30,7 @@ from cudf._lib.types import Interpolation
 
 cimport cudf._lib.cpp.aggregation as libcudf_aggregation
 cimport cudf._lib.cpp.types as libcudf_types
+from cudf._lib.cpp.aggregation cimport underlying_type_t_correlation_type
```
Comment (vyasr) [inline python/cudf/cudf/_lib/aggregation.pyx:33]: This typedef should probably go in `cudf._lib.types` where all the other `underlying_*` typedefs live.

File: python/cudf/cudf/core/groupby/groupby.py [hunk 12]
```diff
@@ -780,6 +784,126 @@ def median(self):
...
+    def corr(self, method="pearson", min_periods=1):
+        """
+        Compute pairwise correlation of columns, excluding NA/null values.
+
+        Parameters
+        ----------
+        method: Method of correlation
+            Pearson: standard correlation coefficient.
+            Kendall, Spearman correlation and callable method
+            not yet supported.
```
Comment (vyasr) [inline python/cudf/cudf/core/groupby/groupby.py:796]:
```suggestion
        method: {"pearson", "kendall", "spearman"} or callable, default "pearson"
            Currently only the pearson correlation coefficient is supported.
```

File: python/cudf/cudf/core/groupby/groupby.py [hunk 13]
```diff
@@ -780,6 +784,126 @@ def median(self):
...
+            )
+
+        # create expanded dataframe consisting all combinations of the
+        # struct columns-pairs to be correlated
+        # i.e (('col1', 'col1'), ('col1', 'col2'), ('col2', 'col2'))
+        _cols = self.grouping.values.columns.tolist()
+
+        if self._by:
+            new_df = cudf.DataFrame._from_data(self.grouping.keys._data)
+            new_df._data.multiindex = False
```
Comment (vyasr) [inline python/cudf/cudf/core/groupby/groupby.py:853]: Is there a reason that this needs to be specified explicitly? Isn't this the default? If it is necessary, is it also required in the else clause?

File: python/cudf/cudf/core/groupby/groupby.py [hunk 14]
```diff
@@ -780,6 +784,126 @@ def median(self):
...
+        else:
+            new_df = cudf.DataFrame._from_data(
+                {}, index=_index_from_data(self.grouping.keys._data)
+            )
+
+        for i in tuple(itertools.combinations_with_replacement(_cols, 2)):
+            new_df._data[i] = cudf.DataFrame._from_data(
+                {"x": self.obj._data[i[0]], "y": self.obj._data[i[1]]}
+            ).to_struct()
```
Comment (vyasr) [inline python/cudf/cudf/core/groupby/groupby.py:862]:
```suggestion
        for x, y in itertools.combinations_with_replacement(_cols, 2):
            new_df._data[(x, y)] = cudf.DataFrame._from_data(
                {"x": self.obj._data[x], "y": self.obj._data[y]}
            ).to_struct()
```

File: python/cudf/cudf/core/groupby/groupby.py [hunk 15]
```diff
@@ -780,6 +784,126 @@ def median(self):
...
+        new_gb = new_df.groupby(by=self._by, level=self._level)
+
+        try:
+            gb_corr = new_gb.agg(lambda x: x.corr(method, min_periods))
+        except RuntimeError as e:
+            if "Unsupported groupby reduction type-agg combination" in str(e):
+                raise TypeError(
+                    "Correlation accepts only numerical column-pairs"
+                ) from e
```
Comment (vyasr) [inline python/cudf/cudf/core/groupby/groupby.py:871]: I don't think we want the `from e` here. We don't want to surface the exception from the Cython API.
```suggestion
                )
```

File: python/cudf/cudf/core/groupby/groupby.py [hunk 16]
```diff
@@ -780,6 +784,126 @@ def median(self):
...
+            else:
+                raise
```
Comment (vyasr) [inline python/cudf/cudf/core/groupby/groupby.py:873]:
```suggestion
            raise
```

File: python/cudf/cudf/core/groupby/groupby.py [hunk 17]
```diff
@@ -780,6 +784,126 @@ def median(self):
...
+            cols_list[i : i + len(_cols)]
+            for i in range(0, len(cols_list), len(_cols))
+        ]
+
+        # interleave: combine the correlation results for each column-pair
+        # into a single column
+        res = cudf.DataFrame()
+        for i, x in zip(cols_split, _cols):
+            ic = gb_corr.loc[:, i].interleave_columns()
+            res[x] = ic
```
Comment (vyasr) [inline python/cudf/cudf/core/groupby/groupby.py:891]:
```suggestion
        res = cudf.DataFrame._from_data({
            x: gb_corr.loc[:, i].interleave_columns()
            for i, x in zip(cols_split, _cols)
        })
```

**Review by vyasr — CHANGES_REQUESTED** [review]
I have some suggestions for improvement before we finalize this, but it looks good!

File: python/cudf/cudf/core/groupby/groupby.py (hunk 13 above)
Comment (isVoid) [reply python/cudf/cudf/core/groupby/groupby.py:853]: I'm not quite sure why in one case the key columns are treated as data columns while in the other as index columns. Is this something might already handled in grouping object?

File: python/cudf/cudf/core/groupby/groupby.py (hunk 14 above)
Comment (isVoid) [reply python/cudf/cudf/core/groupby/groupby.py:862]: Ha! You beat me to it!

**Comment by karthikeyann** [comment]
rerun tests

**Comment by vyasr** [comment]
I think tests are currently running as per our rerun request but the UI is not updating. Let's avoid retriggering tests for this evening and see if anything updates.

File: python/cudf/cudf/core/groupby/groupby.py [hunk 18]
```diff
@@ -70,6 +72,8 @@ def __init__(
         """
         self.obj = obj
         self._as_index = as_index
+        self._by = by
+        self._level = level
```
[the remark this answers is not in the document]
Comment (isVoid) [reply python/cudf/cudf/core/groupby/groupby.py:76]: It turns out this is a pretty special case. While self grouping processes by and level input to get you the key column to pass downward, it doesn't store the original argument. In `corr` we actually want to use the original argument to construct a new groupby object.

File: python/cudf/cudf/core/groupby/groupby.py [hunk 19]
```diff
@@ -780,6 +784,126 @@ def median(self):
...
+        # i.e (('col1', 'col1'), ('col1', 'col2'), ('col2', 'col2'))
+        _cols = self.grouping.values.columns.tolist()
+
+        if self._by:
+            new_df = cudf.DataFrame._from_data(self.grouping.keys._data)
+            new_df._data.multiindex = False
+        else:
+            new_df = cudf.DataFrame._from_data(
+                {}, index=_index_from_data(self.grouping.keys._data)
+            )
```
Comment (isVoid) [inline python/cudf/cudf/core/groupby/groupby.py:857]: After pulling the code and playing with it a bit, I think the process here is basically to make the key columns exists in `new_df` and so when you create a new groupby object it can find those columns. But this is unnecessary, `self.grouping` already processes the `by` and `level` argument to create the key columns you need, which I think this can be simplified to:

```suggestion
        new_df_data = {}
```
TBC (1/3)

File: python/cudf/cudf/core/groupby/groupby.py [hunk 20]
```diff
@@ -780,6 +784,126 @@ def median(self):
...
+
+        for i in tuple(itertools.combinations_with_replacement(_cols, 2)):
+            new_df._data[i] = cudf.DataFrame._from_data(
```
Comment (isVoid) [inline python/cudf/cudf/core/groupby/groupby.py:860]:
```suggestion
            new_df_data[i] = cudf.DataFrame._from_data(
```
TBC (2/3)

File: python/cudf/cudf/core/groupby/groupby.py [hunk 21]
```diff
@@ -780,6 +784,126 @@ def median(self):
...
+                {"x": self.obj._data[i[0]], "y": self.obj._data[i[1]]}
+            ).to_struct()
+        new_gb = new_df.groupby(by=self._by, level=self._level)
```
Comment (isVoid) [inline python/cudf/cudf/core/groupby/groupby.py:863]:
```suggestion
        new_gb = new_gb = cudf.DataFrame._from_data(new_df_data).groupby(by=self.grouping.keys)
```
Fin (3/3)

File: python/cudf/cudf/_lib/aggregation.pyx (hunk 11 above)
Comment (karthikeyann) [reply python/cudf/cudf/_lib/aggregation.pyx:33]: Reason for adding this type to `aggregation.pxd`, is that it is not used at any other place besides aggregations.
didn't want to overload `types.pxd`, which gets included in most files (and hence to avoid extra compilations).

File: python/cudf/cudf/core/groupby/groupby.py [hunk 22]
```diff
@@ -845,19 +845,13 @@ def corr(self, method="pearson", min_periods=1):
...
-            )
+        new_df_data = {}
 
         for i in tuple(itertools.combinations_with_replacement(_cols, 2)):
-            new_df._data[i] = cudf.DataFrame._from_data(
+            new_df_data[i] = cudf.DataFrame._from_data(
                 {"x": self.obj._data[i[0]], "y": self.obj._data[i[1]]}
             ).to_struct()
-        new_gb = new_df.groupby(by=self._by, level=self._level)
+        new_gb = new_gb = cudf.DataFrame._from_data(new_df_data).groupby(by=self.grouping.keys)
```
Comment (shwina) [inline python/cudf/cudf/core/groupby/groupby.py:854]:
```suggestion
        new_gb = cudf.DataFrame._from_data(new_df_data).groupby(by=self.grouping.keys)
```

File: docs/cudf/source/basics/groupby.rst (hunk 9 above)
Comment (skirui-source) [reply docs/cudf/source/basics/groupby.rst:131]: I just liked my example better, but i've added an example using existing dataframe ✔️

File: python/cudf/cudf/_lib/aggregation.pxd (hunk 10 above)
[the remark this answers is not in the document]
Comment (skirui-source) [reply python/cudf/cudf/_lib/aggregation.pxd:1]: Oops that's because I, initially changed the copyright for the wrong aggregation.pxd file.,(now fixed)

File: python/cudf/cudf/core/groupby/groupby.py [hunk 23]
```diff
@@ -777,6 +781,120 @@ def median(self):
...
+                    "Correlation accepts only numerical column-pairs"
+                )
+            raise
+
+        # ensure that column-pair labels are arranged in ascending order
+        cols_list = [
+            (_cols[j], _cols[i]) if i > j else (_cols[i], _cols[j])
+            for j, y in enumerate(_cols)
+            for i, x in enumerate(_cols)
+        ]
```
Comment (isVoid) [inline python/cudf/cudf/core/groupby/groupby.py:869]:
```suggestion
        cols_list = [
            (y, x) if i > j else (x, y)
            for j, y in enumerate(_cols)
            for i, x in enumerate(_cols)
        ]
```

**Review by isVoid — APPROVED** [review]
Just got one small suggestion, else lgtm.

File: python/cudf/cudf/core/groupby/groupby.py (hunk 23 above)
Comment (shwina) [reply python/cudf/cudf/core/groupby/groupby.py:869]: Going to merge in this suggestion if that's OK!

**Comment by skirui-source** [comment]
rerun tests

**Review by mayankanand007 — APPROVED** [review]

File: python/cudf/cudf/core/groupby/groupby.py (hunk 5 above)
Comment (vyasr) [reply python/cudf/cudf/core/groupby/groupby.py:784]: It looks like this isn't done yet. Note that there are two `groupby.rst` files, one in the `basics/` directory and one in the `api_docs` directory. The table in the former has been updated, but not the API link in the latter.

File: python/cudf/cudf/core/groupby/groupby.py [hunk 24]
```diff
@@ -777,6 +781,120 @@ def median(self):
         """Get the column-wise median of the values in each group."""
         return self.agg("median")
 
+    def corr(self, method="pearson", min_periods=1):
+        """
+        Compute pairwise correlation of columns, excluding NA/null values.
+
+        Parameters
+        ----------
+        method: {"pearson" (default), "kendall", "spearman"} or callable
```
Comment (vyasr) [inline python/cudf/cudf/core/groupby/groupby.py:790]: I think this is our normal syntax for this kind of doc
```suggestion
        method: {"pearson", "kendall", "spearman"} or callable, default "pearson"
```

File: python/cudf/cudf/core/groupby/groupby.py [hunk 25]
```diff
@@ -777,6 +781,120 @@ def median(self):
...
+            val2  0.500000  1.000000  1.000000
+            val3  0.500000  1.000000  1.000000
+        b   val1  1.000000  0.385727  0.385727
+            val2  0.385727  1.000000  1.000000
+            val3  0.385727  1.000000  1.000000
+        c   val1  1.000000  0.714575  0.714575
+            val2  0.714575  1.000000  1.000000
+            val3  0.714575  1.000000  1.000000
+
+        """
```
Comment (vyasr) [inline python/cudf/cudf/core/groupby/groupby.py:834]:
```suggestion
        """
```

File: python/cudf/cudf/core/groupby/groupby.py [hunk 26]
```diff
@@ -777,6 +781,120 @@ def median(self):
...
+
+        if not method.lower() in ["pearson"]:
```
Comment (vyasr) [inline python/cudf/cudf/core/groupby/groupby.py:836]: Tuples are preferable for this kind of usage to convey the immutability of the valid set.
```suggestion
        if not method.lower() in ("pearson",):
```

File: python/cudf/cudf/core/groupby/groupby.py [hunk 27]
```diff
@@ -777,6 +781,120 @@ def median(self):
...
+        # to match pandas behavior
+        unsorted_idx = gb_corr.index.repeat(len(_cols))
+        idx_sort_order = unsorted_idx._get_sorted_inds()
+        sorted_idx = unsorted_idx._gather(idx_sort_order)
+        if len(gb_corr):
+            # TO-DO: Should the operation below be done on the CPU instead?
+            sorted_idx._data[None] = as_column(
+                cudf.Series(_cols).tile(len(gb_corr.index))
+            )
+        res.index = _index_from_data(sorted_idx._data)
```
Comment (vyasr) [inline python/cudf/cudf/core/groupby/groupby.py:894]: If you already know that you'll be creating a `MultiIndex` could you call `MultiIndex._from_data` directly?

File: python/cudf/cudf/core/groupby/groupby.py [hunk 28]
```diff
@@ -777,6 +781,120 @@ def median(self):
...
+            raise
+
+        # ensure that column-pair labels are arranged in ascending order
+        cols_list = [
+            (y, x) if i > j else (x, y)
+            for j, y in enumerate(_cols)
+            for i, x in enumerate(_cols)
+        ]
+        cols_split = [
+            cols_list[i : i + len(_cols)]
```
Comment (vyasr) [inline python/cudf/cudf/core/groupby/groupby.py:871]: Minor nit: you're using `len(cols)` in multiple places so you could compute it once when you create `_cols` and cache it.

**Review by vyasr — CHANGES_REQUESTED** [review]
Some minor suggestions then this should be good to go!

**Comment by skirui-source** [comment]
rerun tests

**Review by vyasr — APPROVED** [review]
LGTM!

File: python/cudf/cudf/core/groupby/groupby.py (hunk 27 above)
Comment (shwina) [reply python/cudf/cudf/core/groupby/groupby.py]: I believe it won't always  be a `MultiIndex`, which is why we chose this.  I could be remembering wrong

**Comment by skirui-source** [comment]
@gpucibot merge
